File: android/src/main/java/org/ligi/gobandroid_hd/ui/tsumego/TsumegoGameExtrasFragment.java
package org.ligi.gobandroid_hd.ui.tsumego;

import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.text.Html;
import android.text.TextUtils;
import android.text.method.LinkMovementMethod;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import org.ligi.gobandroid_hd.App;
import org.ligi.gobandroid_hd.R;
import org.ligi.gobandroid_hd.logic.GoGame;
import org.ligi.gobandroid_hd.model.GameProvider;
import org.ligi.gobandroid_hd.ui.fragments.CommentHelper;

import javax.inject.Inject;

import butterknife.Bind;
import butterknife.ButterKnife;

public class TsumegoGameExtrasFragment extends Fragment {

    @Inject
    GameProvider gameProvider;

    @Bind(R.id.tsumego_correct_view)
    TextView correctView;

    @Bind(R.id.tsumego_off_path_view)
    View OffPathView;

    @Bind(R.id.game_comment)
    TextView commentView;

    private boolean off_path_visible = false, correct_visible = false;

    private void updateUI() {
        if (OffPathView == null || correctView == null || getActivity() == null) { // views not yet created
            return; // will come back later
        }

        final GoGame game = gameProvider.get();

        getActivity().runOnUiThread(new Runnable() {
            @Override
            public void run() {

                OffPathView.setVisibility(off_path_visible ? TextView.VISIBLE : TextView.GONE);

                if (correct_visible) {
                    correctView.setVisibility(View.VISIBLE);
                    final String optionalNextTsumegoURLString = NextTsumegoFileFinder.calcNextTsumego(game.getMetaData()
                                                                                                          .getFileName()
                                                                                                          .replaceFirst("file://", ""));

                    if (optionalNextTsumegoURLString != null) {

                        correctView.setMovementMethod(LinkMovementMethod.getInstance());

                        final String text = getString(R.string.tsumego_correct) +
                                            " <a href='tsumego://" +
                                            optionalNextTsumegoURLString +
                                            "'>" +
                                            getString(R.string.next_tsumego) +
                                            "</a>";
                        correctView.setText(Html.fromHtml(text));
                    } else {
                        correctView.setText(getString(R.string.correct_but_no_more_tsumegos));
                    }
                } else {
                    correctView.setVisibility(View.GONE);
                }

                // the 10 is a bit of a magic number - just want to show comments that
                // have extras here to prevent double commentView written - but sometimes
                // there is more info in the commentView
                if (!correct_visible && game.getActMove().getComment().length() > 10) {
                    commentView.setVisibility(View.VISIBLE);
                    commentView.setText(game.getActMove().getComment());
                    if (!TextUtils.isEmpty(game.getActMove().getComment())) {
                        CommentHelper.linkifyCommentTextView(commentView);
                    }
                } else {
                    commentView.setVisibility(View.GONE);
                }
            }
        });

    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        App.component().inject(this);

        final View res = inflater.inflate(R.layout.game_extra_tsumego, container, false);

        ButterKnife.bind(this, res);

        updateUI();
        return res;
    }

    public void setOffPathVisibility(boolean visible) {
        off_path_visible = visible;
        updateUI();
    }

    public void setCorrectVisibility(boolean visible) {
        correct_visible = visible;
        updateUI();
    }

}

File: android/src/main/java/org/ligi/gobandroid_hd/ui/tsumego/TsumegoGameExtrasFragment.kt
package org.ligi.gobandroid_hd.ui.tsumego

import android.os.Bundle
import androidx.fragment.app.Fragment
import android.text.TextUtils
import android.text.method.LinkMovementMethod
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import com.github.salomonbrys.kodein.instance
import com.github.salomonbrys.kodein.lazy
import kotlinx.android.synthetic.main.game_extra_tsumego.*
import org.ligi.compat.HtmlCompat
import org.ligi.gobandroid_hd.App
import org.ligi.gobandroid_hd.R
import org.ligi.gobandroid_hd.model.GameProvider
import org.ligi.gobandroid_hd.ui.go_terminology.GoTerminologyViewActivity

class TsumegoGameExtrasFragment : Fragment() {

    internal val gameProvider: GameProvider  by App.kodein.lazy.instance()


    private var off_path_visible = false
    private var correct_visible = false

    private fun updateUI() {
        if (tsumego_off_path_view == null || tsumego_correct_view == null || activity == null) { // views not yet created
            return  // will come back later
        }

        val game = gameProvider.get()

        requireActivity().runOnUiThread {
            tsumego_off_path_view.visibility = if (off_path_visible) TextView.VISIBLE else TextView.GONE

            if (correct_visible) {
                tsumego_correct_view!!.visibility = View.VISIBLE
                val optionalNextTsumegoURLString = NextTsumegoFileFinder.calcNextTsumego(game.metaData
                        .fileName
                        .replaceFirst("file://".toRegex(), ""))

                if (optionalNextTsumegoURLString != null) {

                    tsumego_correct_view!!.movementMethod = LinkMovementMethod.getInstance()

                    val text = getString(R.string.tsumego_correct) +
                            " <a href='tsumego://" +
                            optionalNextTsumegoURLString +
                            "'>" +
                            getString(R.string.next_tsumego) +
                            "</a>"
                    tsumego_correct_view!!.text = HtmlCompat.fromHtml(text)
                } else {
                    tsumego_correct_view.text = getString(R.string.correct_but_no_more_tsumegos)
                }
            } else {
                tsumego_correct_view.visibility = View.GONE
            }

            // the 10 is a bit of a magic number - just want to show comments that
            // have extras here to prevent double commentView written - but sometimes
            // there is more info in the commentView
            if (!correct_visible && game.actMove.comment.length > 10) {
                game_comment!!.visibility = View.VISIBLE
                game_comment!!.text = game.actMove.comment
                if (!TextUtils.isEmpty(game.actMove.comment)) {
                    GoTerminologyViewActivity.linkifyTextView(game_comment!!)
                }
            } else {
                game_comment!!.visibility = View.GONE
            }
        }

    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val res = inflater.inflate(R.layout.game_extra_tsumego, container, false)

        updateUI()
        return res
    }

    fun setOffPathVisibility(visible: Boolean) {
        off_path_visible = visible
        updateUI()
    }

    fun setCorrectVisibility(visible: Boolean) {
        correct_visible = visible
        updateUI()
    }

}
